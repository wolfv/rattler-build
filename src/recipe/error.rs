use std::{fmt, str::ParseBoolError};

use miette::{Diagnostic, SourceOffset, SourceSpan};
use thiserror::Error;

/// The Error type for the first stage of the recipe parser.
///
/// This type was designed to be compatible with [`miette`], and its [`Diagnostic`] trait.
#[derive(Debug, Error, Diagnostic)]
#[error("Parsing: {kind}")]
pub struct ParsingError {
    // Source string of the recipe.
    #[source_code]
    pub src: String,

    /// Offset in chars of the error.
    #[label("{}", label.unwrap_or("here"))]
    pub span: SourceSpan,

    /// Label text for this span. Defaults to `"here"`.
    pub label: Option<&'static str>,

    /// Suggestion for fixing the parser error.
    #[help]
    pub help: Option<&'static str>,

    // Specific error kind for the error.
    pub kind: ErrorKind,
}

impl ParsingError {
    pub fn from_partial(src: &str, err: PartialParsingError) -> Self {
        Self {
            src: src.to_owned(),
            span: marker_span_to_span(src, err.span),
            label: err.label,
            help: err.help,
            kind: err.kind,
        }
    }

    pub fn kind(&self) -> &ErrorKind {
        &self.kind
    }
}

/// Type that represents the kind of error that can happen in the first stage of the recipe parser.
#[derive(Debug, Error, Diagnostic)]
#[non_exhaustive]
pub enum ErrorKind {
    /// Error while parsing YAML.
    #[diagnostic(code(error::stage1::yaml_parsing))]
    YamlParsing(#[from] marked_yaml::LoadError),

    /// Error when expected mapping but got something else.
    #[diagnostic(code(error::stage1::expected_mapping))]
    ExpectedMapping,

    /// Error when expected scalar but got something else.
    #[diagnostic(code(error::stage1::expected_scalar))]
    ExpectedScalar,

    /// Error when expected sequence but got something else.
    #[diagnostic(code(error::stage1::expected_sequence))]
    ExpectedSequence,

    /// Error when if-selector condition is not a scalar.
    #[diagnostic(code(error::stage1::if_selector_condition_not_scalar))]
    IfSelectorConditionNotScalar,

    /// Error when if selector is missing a `then` field.
    #[diagnostic(code(error::stage1::if_selector_missing_then))]
    IfSelectorMissingThen,

    /// Error rendering a Jinja expression.
    #[diagnostic(code(error::stage2::jinja_rendering))]
    JinjaRendering(#[from] minijinja::Error),

    /// Error processing the condition of a if-selector.
    #[diagnostic(code(error::stage2::if_selector_condition_not_bool))]
    IfSelectorConditionNotBool(#[from] ParseBoolError),

    /// Generic unspecified error. If this is returned, the call site should
    /// be annotated with context, if possible.
    #[diagnostic(code(error::other))]
    Other,
}

/// Partial error type, almost the same as the [`ParsingError`] but without the source string.
///
/// This is to use on the context where you want to produce a [`ParsingError`] but you don't have
/// the source string, or including the source string would involve more complexity to handle. Like
/// leveraging traits, simple conversions, etc.
///
/// Examples of this is [`Node`](crate::recipe::stage1::Node) to implement [`TryFrom`] for some
/// types.
#[derive(Debug, Error)]
#[error("Parsing: {kind}")]
pub struct PartialParsingError {
    /// Offset in chars of the error.
    pub span: marked_yaml::Span,

    /// Label text for this span. Defaults to `"here"`.
    pub label: Option<&'static str>,

    /// Suggestion for fixing the parser error.
    pub help: Option<&'static str>,

    // Specific error kind for the error.
    pub kind: ErrorKind,
}

// Implement Display for ErrorKind manually bacause [`marked_yaml::LoadError`] does not implement
// the way we want it.
// CAUTION: Because of this impl, we cannot use `#[error()]` on the enum.
impl fmt::Display for ErrorKind {
    #[allow(deprecated)] // for e.description()
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        use marked_yaml::LoadError;
        use std::error::Error;

        match self {
            ErrorKind::YamlParsing(err) => {
                write!(f, "Failed to parse YAML: ")?;
                match err {
                    LoadError::TopLevelMustBeMapping(_) => {
                        write!(f, "Top level must be a mapping.")
                    }
                    LoadError::UnexpectedAnchor(_) => write!(f, "Unexpected definition of anchor."),
                    LoadError::MappingKeyMustBeScalar(_) => {
                        write!(f, "Keys in mappings must be scalar.")
                    }
                    LoadError::UnexpectedTag(_) => write!(f, "Unexpected use of YAML tag."),
                    LoadError::ScanError(_, e) => {
                        // e.description() is deprecated but it's the only way to get
                        // the exact info we want out of yaml-rust

                        write!(f, "{}", e.description())
                    }
                }
            }
            ErrorKind::ExpectedMapping => write!(f, "Expected a mapping."),
            ErrorKind::ExpectedScalar => write!(f, "Expected a scalar value."),
            ErrorKind::ExpectedSequence => write!(f, "Expected a sequence."),
            ErrorKind::IfSelectorConditionNotScalar => {
                write!(f, "Condition in `if` selector must be a scalar.")
            }
            ErrorKind::IfSelectorMissingThen => {
                write!(f, "Missing `then` field in the `if` selector.")
            }
            ErrorKind::JinjaRendering(err) => {
                write!(f, "Failed to render Jinja expression: {}", err.kind())
            }
            ErrorKind::IfSelectorConditionNotBool(err) => {
                write!(f, "Condition in `if` selector must be a boolean: {}", err)
            }
            ErrorKind::Other => write!(f, "An unspecified error occurred."),
        }
    }
}

/// Macro to facilitate the creation of [`Error`]s.
#[macro_export]
#[doc(hidden)]
macro_rules! _error {
    ($src:expr, $span:expr, $kind:expr $(,)?) => {{
        $crate::recipe::error::ParsingError {
            src: $src.to_owned(),
            span: $span,
            label: None,
            help: None,
            kind: $kind,
        }
    }};
    ($src:expr, $span:expr, $kind:expr, label = $label:expr $(,)?) => {{
        $crate::recipe::error::ParsingError {
            src: $src.to_owned(),
            span: $span,
            label: Some($label),
            help: None,
            kind: $kind,
        }
    }};
    ($src:expr, $span:expr, $kind:expr, help = $help:expr $(,)?) => {{
        $crate::recipe::error::ParsingError {
            src: $src.to_owned(),
            span: $span,
            label: None,
            help: Some($help),
            kind: $kind,
        }
    }};
    ($src:expr, $span:expr, $kind:expr, label = $label:expr, help = $help:expr $(,)?) => {{
        $crate::_error!($src, $span, $kind, $label, $help)
    }};
    ($src:expr, $span:expr, $kind:expr, help = $help:expr, label = $label:expr $(,)?) => {{
        $crate::_error!($src, $span, $kind, $label, $help)
    }};
    ($src:expr, $span:expr, $kind:expr, $label:expr, $help:expr $(,)?) => {{
        $crate::recipe::error::ParsingError {
            src: $src.to_owned(),
            span: $span,
            label: Some($label),
            help: Some($help),
            kind: $kind,
        }
    }};
}

/// Macro to facilitate the creation of [`Error`]s.
#[macro_export]
#[doc(hidden)]
macro_rules! _partialerror {
    ($span:expr, $kind:expr $(,)?) => {{
        $crate::recipe::error::PartialParsingError {
            span: $span,
            label: None,
            help: None,
            kind: $kind,
        }
    }};
    ($span:expr, $kind:expr, label = $label:expr $(,)?) => {{
        $crate::recipe::error::PartialParsingError {
            span: $span,
            label: Some($label),
            help: None,
            kind: $kind,
        }
    }};
    ($span:expr, $kind:expr, help = $help:expr $(,)?) => {{
        $crate::recipe::error::PartialParsingError {
            span: $span,
            label: None,
            help: Some($help),
            kind: $kind,
        }
    }};
    ($span:expr, $kind:expr, label = $label:expr, help = $help:expr $(,)?) => {{
        $crate::_error!($src, $span, $kind, $label, $help)
    }};
    ($span:expr, $kind:expr, help = $help:expr, label = $label:expr $(,)?) => {{
        $crate::_error!($src, $span, $kind, $label, $help)
    }};
    ($span:expr, $kind:expr, $label:expr, $help:expr $(,)?) => {{
        $crate::recipe::error::PartialParsingError {
            span: $span,
            label: Some($label),
            help: Some($help),
            kind: $kind,
        }
    }};
}

/// Error handler for [`marked_yaml::LoadError`].
pub(super) fn load_error_handler(src: &str, err: marked_yaml::LoadError) -> ParsingError {
    _error!(
        src,
        marker_to_span(src, marker(&err)),
        ErrorKind::YamlParsing(err),
        label = match err {
            marked_yaml::LoadError::TopLevelMustBeMapping(_) => "expected a mapping here",
            marked_yaml::LoadError::UnexpectedAnchor(_) => "unexpected anchor here",
            marked_yaml::LoadError::UnexpectedTag(_) => "unexpected tag here",
            _ => "here",
        }
    )
}

/// Convert a [`marked_yaml::Marker`] to a [`SourceSpan`].
pub(super) fn marker_to_span(src: &str, mark: marked_yaml::Marker) -> SourceSpan {
    let start = SourceOffset::from_location(src, mark.line(), mark.column());

    SourceSpan::new(
        start,
        SourceOffset::from(find_length(src, start)), //::from_location(src, mark.line(), mark.column()),
    )
}

/// Get the [`marked_yaml::Marker`] from a [`marked_yaml::LoadError`].
pub(super) fn marker(err: &marked_yaml::LoadError) -> marked_yaml::Marker {
    use marked_yaml::LoadError::*;
    match err {
        TopLevelMustBeMapping(m) => *m,
        UnexpectedAnchor(m) => *m,
        MappingKeyMustBeScalar(m) => *m,
        UnexpectedTag(m) => *m,
        ScanError(m, _) => *m,
    }
}

/// Convert a [`marked_yaml::Span`] to a [`SourceSpan`].
pub(super) fn marker_span_to_span(src: &str, span: marked_yaml::Span) -> SourceSpan {
    let marked_start = span
        .start()
        .copied()
        .unwrap_or_else(|| marked_yaml::Marker::new(usize::MAX, usize::MAX, usize::MAX));
    let marked_end = span.end().copied();

    let start = SourceOffset::from_location(src, marked_start.line(), marked_start.column());

    let length = match marked_end {
        Some(end) => {
            let end = SourceOffset::from_location(src, end.line() - 1, end.column());
            end.offset() - start.offset() - 1
        }
        None => {
            let l = find_length(src, start);
            if l == 0 {
                1
            } else {
                l
            }
        }
    };

    SourceSpan::new(start, SourceOffset::from(length))
}

#[allow(dead_code)]
pub(super) fn marker_to_offset(src: &str, mark: marked_yaml::Marker) -> SourceOffset {
    SourceOffset::from_location(src, mark.line(), mark.column())
}

/// Find the length of the token string starting at the `start` [`SourceOffset`].
pub(super) fn find_length(src: &str, start: SourceOffset) -> usize {
    let start = start.offset();
    let mut end = 0;

    for (i, c) in src[start..].char_indices() {
        if c.is_whitespace() {
            end += i;
            break;
        }
    }

    end
}

/// Asserts a [`miette::Report`] snapshot.
///
/// The value needs to implement the `fmt::Debug` trait.  This is useful for
/// simple values that do not implement the `Serialize` trait but does not
/// permit redactions.
///
/// Debug is called with `"{:#?}"`, which means this uses pretty-print.
///
/// The snapshot name is optional.
#[cfg_attr(test, macro_export)]
#[allow(unused_macros)]
macro_rules! assert_miette_snapshot {
    ($value:expr, @$snapshot:literal) => {{
        let value: &::miette::Report = &$value;
        let value = format!("{:?}", $value);
        ::insta::assert_snapshot!(value, stringify!($value), @$snapshot);
    }};
    ($name:expr, $value:expr) => {{
        let value: &::miette::Report = &$value;
        let value = format!("{:?}", value);
        ::insta::assert_snapshot!(Some($name), value, stringify!($value));
    }};
    ($value:expr) => {{
        let value: &::miette::Report = &$value;
        let value = format!("{:?}", value);
        ::insta::assert_snapshot!(::insta::_macro_support::AutoName, value, stringify!($value));
    }};
}

#[cfg(test)]
mod tests {

    use crate::assert_miette_snapshot;
    use crate::recipe::stage1::RawRecipe;

    #[test]
    fn miette_output() {
        fn test() -> miette::Result<RawRecipe> {
            let fault_yaml = r#"
            context:
                - a
                - b
            package:
                name: test
                version: 0.1.0
            "#;
            let res = RawRecipe::from_yaml(fault_yaml)?;
            Ok(res)
        }

        let res = test();

        assert_miette_snapshot!("{:?}", res.unwrap_err());
    }
}
