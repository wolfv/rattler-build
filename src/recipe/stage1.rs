//! First stage of the recipe pipeline.
//!

use linked_hash_map::LinkedHashMap;

use super::error::{markerspan2span, Error, ErrorKind};

pub mod node;
pub use node::Node;

use crate::_error;

use self::node::{MappingNode, ScalarNode};

/// This is the raw reprentation of a recipe, without any minijinja processing done.
///
/// This is the type that is used to parse the recipe file in the first stage, and only validates
/// the existance of the root keys and required keys but not their values (that can be jinja syntax).
#[derive(Debug, Clone, PartialEq)]
pub struct RawRecipe {
    pub(crate) context: LinkedHashMap<ScalarNode, ScalarNode>,
    pub(crate) package: Package,
    pub(crate) source: Source,
    pub(crate) build: Build,
    pub(crate) requirements: Option<Requirements>,
    pub(crate) test: Test,
    pub(crate) about: Option<About>,
    pub(crate) extra: Extra,
}

impl RawRecipe {
    /// Parse a recipe from a YAML string.
    pub fn from_yaml(yaml: &str) -> Result<Self, Error> {
        let yaml_root = marked_yaml::parse_yaml(0, yaml)
            .map_err(|err| super::error::load_error_handler(yaml, err))?;

        let yaml_root = Node::try_from(yaml_root)
            .map_err(|err| _error!(yaml, markerspan2span(yaml, err.span), err.kind))?;

        let root_map = yaml_root.as_mapping().expect("top level must be a mapping");

        let mut context = LinkedHashMap::new();
        let mut package = Package {
            name: ScalarNode::from(""),
            version: ScalarNode::from(""),
        };
        let mut source = Source::default();
        let mut build = Build::default();
        let mut requirements = None;
        let mut test = Test::default();
        let mut about = None;
        let mut extra = Extra::default();

        for (key, value) in root_map.iter() {
            let key = key.as_str();

            match key {
                "context" => {
                    let context_node = value;
                    let context_span = markerspan2span(yaml, *context_node.span());

                    let context_mapping = context_node.as_mapping().ok_or({
                        _error!(
                            yaml,
                            context_span,
                            ErrorKind::ExpectedMapping,
                            label = "expected a mapping here",
                        )
                    })?;

                    for (key, value) in context_mapping.iter() {
                        let value = value.as_scalar().ok_or_else(|| {
                            _error!(
                                yaml,
                                markerspan2span(yaml, *value.span()),
                                ErrorKind::ExpectedScalar,
                                label = "expected a scalar value here",
                            )
                        })?;

                        context.insert(key.clone(), value.clone());
                    }
                }
                "package" => {
                    if let Some(package_node) = value.as_mapping() {
                        let package_span = markerspan2span(yaml, *package_node.span());

                        let mut name = None;
                        let mut version = None;

                        for (key, value) in package_node.iter() {
                            let key = key.as_str();

                            match key {
                                "name" => {
                                    if let Some(name_node) = value.as_scalar() {
                                        name = Some(name_node.clone());
                                    } else {
                                        return Err(_error!(
                                            yaml,
                                            markerspan2span(yaml, *value.span()),
                                            ErrorKind::ExpectedScalar,
                                            label = "expected a scalar value here",
                                        ));
                                    }
                                }
                                "version" => {
                                    if let Some(version_node) = value.as_scalar() {
                                        version = Some(version_node.clone());
                                    } else {
                                        return Err(_error!(
                                            yaml,
                                            markerspan2span(yaml, *value.span()),
                                            ErrorKind::ExpectedScalar,
                                            label = "expected a scalar value here",
                                        ));
                                    }
                                }
                                _ => {
                                    return Err(_error!(
                                        yaml,
                                        markerspan2span(yaml, *value.span()),
                                        ErrorKind::Other,
                                        label = "unexpected key",
                                        help = "expected one of `name` or `version`"
                                    ));
                                }
                            }
                        }

                        let name = name.ok_or_else(|| {
                            _error!(
                                yaml,
                                package_span,
                                ErrorKind::Other,
                                label = "missing key `name`",
                            )
                        })?;

                        let version = version.ok_or_else(|| {
                            _error!(
                                yaml,
                                package_span,
                                ErrorKind::Other,
                                label = "missing key `version`",
                            )
                        })?;

                        package = Package { name, version };
                    } else {
                        return Err(_error!(
                            yaml,
                            markerspan2span(yaml, *value.span()),
                            ErrorKind::ExpectedMapping,
                            label = "expected a mapping here",
                        ));
                    }
                }
                "source" => source.node = Some(value.clone()),
                "build" => build.node = value.as_mapping().cloned(),
                "requirements" => {
                    if let Some(requirements_node) = value.as_mapping() {
                        let requirements_span = markerspan2span(yaml, *requirements_node.span());

                        let mut req = Requirements::default();

                        for (key, value) in requirements_node.iter() {
                            let key = key.as_str();

                            match key {
                                "build" => req.build = Some(value.clone()),
                                "host" => req.host = Some(value.clone()),
                                "run" => req.run = Some(value.clone()),
                                "run_constrained" => req.run_constrained = Some(value.clone()),
                                _ => {
                                    return Err(_error!(
                                        yaml,
                                        requirements_span,
                                        ErrorKind::Other,
                                        label = "unexpected key",
                                        help = "expected one of `build`, `host`, `run` or `run_constrained`"
                                    ));
                                }
                            }
                        }

                        requirements = Some(req);
                    } else {
                        return Err(_error!(
                            yaml,
                            markerspan2span(yaml, *value.span()),
                            ErrorKind::ExpectedMapping,
                            label = "expected a mapping here",
                        ));
                    }
                }
                "test" => test.node = Some(value.clone()),
                "about" => {
                    if let Some(about_node) = value.as_mapping() {
                        let about_span = markerspan2span(yaml, *about_node.span());

                        let mut ab = About::default();

                        for (key, value) in about_node.iter() {
                            let key = key.as_str();

                            match key {
                                "homepage" | "home" => ab.homepage = Some(value.clone()),
                                "repository" | "dev_url" => {
                                    if let Some(repository_node) = value.as_scalar() {
                                        ab.repository = Some(repository_node.clone());
                                    } else {
                                        return Err(_error!(
                                            yaml,
                                            markerspan2span(yaml, *value.span()),
                                            ErrorKind::ExpectedScalar,
                                            label = "expected a scalar value here",
                                        ));
                                    }
                                }
                                "documentation" | "doc_url" => {
                                    if let Some(documentation_node) = value.as_scalar() {
                                        ab.documentation = Some(documentation_node.clone());
                                    } else {
                                        return Err(_error!(
                                            yaml,
                                            markerspan2span(yaml, *value.span()),
                                            ErrorKind::ExpectedScalar,
                                            label = "expected a scalar value here",
                                        ));
                                    }
                                }
                                "license" => {
                                    if let Some(license_node) = value.as_scalar() {
                                        ab.license = Some(license_node.clone());
                                    } else {
                                        return Err(_error!(
                                            yaml,
                                            markerspan2span(yaml, *value.span()),
                                            ErrorKind::ExpectedScalar,
                                            label = "expected a scalar value here",
                                        ));
                                    }
                                }
                                "license_family" => {
                                    if let Some(license_family_node) = value.as_scalar() {
                                        ab.license_family = Some(license_family_node.clone());
                                    } else {
                                        return Err(_error!(
                                            yaml,
                                            markerspan2span(yaml, *value.span()),
                                            ErrorKind::ExpectedScalar,
                                            label = "expected a scalar value here",
                                        ));
                                    }
                                }
                                "license_file" => ab.license_file = Some(value.clone()),
                                "license_url" => {
                                    if let Some(license_url_node) = value.as_scalar() {
                                        ab.license_url = Some(license_url_node.clone());
                                    } else {
                                        return Err(_error!(
                                            yaml,
                                            markerspan2span(yaml, *value.span()),
                                            ErrorKind::ExpectedScalar,
                                            label = "expected a scalar value here",
                                        ));
                                    }
                                }
                                "summary" => {
                                    if let Some(summary_node) = value.as_scalar() {
                                        ab.summary = Some(summary_node.clone());
                                    } else {
                                        return Err(_error!(
                                            yaml,
                                            markerspan2span(yaml, *value.span()),
                                            ErrorKind::ExpectedScalar,
                                            label = "expected a scalar value here",
                                        ));
                                    }
                                }
                                "description" => {
                                    if let Some(description_node) = value.as_scalar() {
                                        ab.description = Some(description_node.clone());
                                    } else {
                                        return Err(_error!(
                                            yaml,
                                            markerspan2span(yaml, *value.span()),
                                            ErrorKind::ExpectedScalar,
                                            label = "expected a scalar value here",
                                        ));
                                    }
                                }
                                "prelink_message" => {
                                    if let Some(prelink_message_node) = value.as_scalar() {
                                        ab.prelink_message = Some(prelink_message_node.clone());
                                    } else {
                                        return Err(_error!(
                                            yaml,
                                            markerspan2span(yaml, *value.span()),
                                            ErrorKind::ExpectedScalar,
                                            label = "expected a scalar value here",
                                        ));
                                    }
                                }
                                _ => {
                                    return Err(_error!(
                                        yaml,
                                        markerspan2span(yaml, *value.span()),
                                        ErrorKind::Other,
                                        label = "unexpected key",
                                        help = "expected one of `homepage`, `repository`, `documentation`, `license`, `license_file`, `license_url`, `summary`, `description` or `prelink_message`"
                                    ));
                                }
                            }
                        }

                        about = Some(ab);
                    } else {
                        return Err(_error!(
                            yaml,
                            markerspan2span(yaml, *value.span()),
                            ErrorKind::ExpectedMapping,
                            label = "expected a mapping here",
                        ));
                    }
                }
                "extra" => extra.node = Some(value.clone()),
                _ => {
                    return Err(_error!(
                        yaml,
                        markerspan2span(yaml, *value.span()),
                        ErrorKind::Other,
                        label = "unexpected key",
                        help = "expected one of `context`, `package`, `source`, `build`, `requirements`, `test`, `about` or `extra`"
                    ));
                }
            }
        }

        Ok(Self {
            context,
            package,
            source,
            build,
            requirements,
            test,
            about,
            extra,
        })
    }
}

/// A package with name and version
#[derive(Debug, Clone, PartialEq)]
pub struct Package {
    /// The package name
    pub(crate) name: ScalarNode,
    /// The package version
    pub(crate) version: ScalarNode,
}

/// A source of a package
///
/// There are many possibilities for this field that cannot be semantically checked
/// in the first stage. It is optional, there is no required fields, allows for if-selector,
/// certain fields only occours with another field.
#[derive(Debug, Clone, Default, PartialEq)]
pub struct Source {
    pub(crate) node: Option<Node>,
}

/// A build of a package
///
/// There are many possibilities for this field that cannot be semantically checked
/// in the first stage. It is optional, there is no required fields, allows for if-selector,
/// certain fields only occours with another field.
#[derive(Debug, Default, Clone, PartialEq)]
pub struct Build {
    pub(crate) node: Option<MappingNode>,
}

/// A requirements of a package (dependencies)
#[derive(Debug, Default, Clone, PartialEq)]
pub struct Requirements {
    pub(crate) build: Option<Node>,
    pub(crate) host: Option<Node>,
    pub(crate) run: Option<Node>,
    pub(crate) run_constrained: Option<Node>,
}

/// A tests of a package
///
/// There are many possibilities for this field that cannot be semantically checked
/// in the first stage. It is optional, there is no required fields, allows for if-selector,
/// certain fields only occours with another field.
#[derive(Debug, Default, Clone, PartialEq)]
pub struct Test {
    pub(crate) node: Option<Node>,
}

/// A package about information
#[derive(Debug, Default, Clone, PartialEq)]
pub struct About {
    pub(crate) homepage: Option<Node>,
    pub(crate) repository: Option<ScalarNode>,
    pub(crate) documentation: Option<ScalarNode>,
    pub(crate) license: Option<ScalarNode>,
    pub(crate) license_family: Option<ScalarNode>,
    pub(crate) license_file: Option<Node>,
    pub(crate) license_url: Option<ScalarNode>,
    pub(crate) summary: Option<ScalarNode>,
    pub(crate) description: Option<ScalarNode>,
    pub(crate) prelink_message: Option<ScalarNode>,
}

/// A tests of a package
///
/// There are many possibilities for this field that cannot be semantically checked
/// in the first stage. It is optional, there is no required fields, allows for if-selector,
/// certain fields only occours with another field.
#[derive(Debug, Default, Clone, PartialEq)]
pub struct Extra {
    pub(crate) node: Option<Node>,
}

#[cfg(test)]
mod tests {

    use insta::assert_debug_snapshot;

    use super::*;

    #[test]
    fn test_parse() {
        let raw_recipe = include_str!("stage1/testfiles/xtensor_recipe.yaml");
        let raw_recipe = RawRecipe::from_yaml(raw_recipe);
        assert!(raw_recipe.is_ok());

        let raw_recipe = raw_recipe.unwrap();

        assert_debug_snapshot!(raw_recipe);
    }
}
