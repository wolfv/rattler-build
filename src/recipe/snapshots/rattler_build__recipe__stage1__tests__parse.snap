---
source: src/recipe/stage1.rs
expression: raw_recipe
---
RawRecipe {
    context: {
        ScalarNode {
            value: "name",
        }: ScalarNode {
            value: "xtensor",
        },
        ScalarNode {
            value: "version",
        }: ScalarNode {
            value: "0.24.6",
        },
        ScalarNode {
            value: "sha256",
        }: ScalarNode {
            value: "f87259b51aabafdd1183947747edfff4cff75d55375334f2e81cee6dc68ef655",
        },
    },
    package: Package {
        name: ScalarNode {
            value: "${{ name|lower }}",
        },
        version: ScalarNode {
            value: "${{ version }}",
        },
    },
    source: Source {
        node: Some(
            Mapping(
                MappingNode {
                    value: {
                        ScalarNode {
                            value: "url",
                        }: Scalar(
                            ScalarNode {
                                value: "https://github.com/xtensor-stack/xtensor/archive/${{ version }}.tar.gz",
                            },
                        ),
                        ScalarNode {
                            value: "sha256",
                        }: Scalar(
                            ScalarNode {
                                value: "${{ sha256 }}",
                            },
                        ),
                    },
                },
            ),
        ),
    },
    build: Build {
        node: Some(
            MappingNode {
                value: {
                    ScalarNode {
                        value: "number",
                    }: Scalar(
                        ScalarNode {
                            value: "0",
                        },
                    ),
                    ScalarNode {
                        value: "skip",
                    }: Scalar(
                        ScalarNode {
                            value: "${{ true if (win and vc14) }}",
                        },
                    ),
                },
            },
        ),
    },
    requirements: Some(
        Requirements {
            build: Some(
                Sequence(
                    SequenceNode {
                        value: [
                            Simple(
                                Scalar(
                                    ScalarNode {
                                        value: "${{ compiler('cxx') }}",
                                    },
                                ),
                            ),
                            Simple(
                                Scalar(
                                    ScalarNode {
                                        value: "cmake",
                                    },
                                ),
                            ),
                            Conditional(
                                IfSelector {
                                    cond: ScalarNode {
                                        value: "unix",
                                    },
                                    then: Scalar(
                                        ScalarNode {
                                            value: "make",
                                        },
                                    ),
                                    otherwise: None,
                                },
                            ),
                        ],
                    },
                ),
            ),
            host: Some(
                Sequence(
                    SequenceNode {
                        value: [
                            Simple(
                                Scalar(
                                    ScalarNode {
                                        value: "xtl >=0.7,<0.8",
                                    },
                                ),
                            ),
                        ],
                    },
                ),
            ),
            run: Some(
                Sequence(
                    SequenceNode {
                        value: [
                            Simple(
                                Scalar(
                                    ScalarNode {
                                        value: "xtl >=0.7,<0.8",
                                    },
                                ),
                            ),
                        ],
                    },
                ),
            ),
            run_constrained: Some(
                Sequence(
                    SequenceNode {
                        value: [
                            Simple(
                                Scalar(
                                    ScalarNode {
                                        value: "xsimd >=8.0.3,<10",
                                    },
                                ),
                            ),
                        ],
                    },
                ),
            ),
        },
    ),
    test: Test {
        node: Some(
            Mapping(
                MappingNode {
                    value: {
                        ScalarNode {
                            value: "commands",
                        }: Sequence(
                            SequenceNode {
                                value: [
                                    Conditional(
                                        IfSelector {
                                            cond: ScalarNode {
                                                value: "unix",
                                            },
                                            then: Sequence(
                                                SequenceNode {
                                                    value: [
                                                        Simple(
                                                            Scalar(
                                                                ScalarNode {
                                                                    value: "test -d ${PREFIX}/include/xtensor",
                                                                },
                                                            ),
                                                        ),
                                                        Simple(
                                                            Scalar(
                                                                ScalarNode {
                                                                    value: "test -f ${PREFIX}/include/xtensor/xarray.hpp",
                                                                },
                                                            ),
                                                        ),
                                                        Simple(
                                                            Scalar(
                                                                ScalarNode {
                                                                    value: "test -f ${PREFIX}/share/cmake/xtensor/xtensorConfig.cmake",
                                                                },
                                                            ),
                                                        ),
                                                        Simple(
                                                            Scalar(
                                                                ScalarNode {
                                                                    value: "test -f ${PREFIX}/share/cmake/xtensor/xtensorConfigVersion.cmake",
                                                                },
                                                            ),
                                                        ),
                                                    ],
                                                },
                                            ),
                                            otherwise: None,
                                        },
                                    ),
                                    Conditional(
                                        IfSelector {
                                            cond: ScalarNode {
                                                value: "win",
                                            },
                                            then: Sequence(
                                                SequenceNode {
                                                    value: [
                                                        Simple(
                                                            Scalar(
                                                                ScalarNode {
                                                                    value: "if not exist %LIBRARY_PREFIX%\\include\\xtensor\\xarray.hpp (exit 1)",
                                                                },
                                                            ),
                                                        ),
                                                        Simple(
                                                            Scalar(
                                                                ScalarNode {
                                                                    value: "if not exist %LIBRARY_PREFIX%\\share\\cmake\\xtensor\\xtensorConfig.cmake (exit 1)",
                                                                },
                                                            ),
                                                        ),
                                                        Simple(
                                                            Scalar(
                                                                ScalarNode {
                                                                    value: "if not exist %LIBRARY_PREFIX%\\share\\cmake\\xtensor\\xtensorConfigVersion.cmake (exit 1)",
                                                                },
                                                            ),
                                                        ),
                                                    ],
                                                },
                                            ),
                                            otherwise: None,
                                        },
                                    ),
                                ],
                            },
                        ),
                    },
                },
            ),
        ),
    },
    about: Some(
        About {
            homepage: Some(
                Scalar(
                    ScalarNode {
                        value: "https://github.com/xtensor-stack/xtensor",
                    },
                ),
            ),
            repository: Some(
                ScalarNode {
                    value: "https://github.com/xtensor-stack/xtensor",
                },
            ),
            documentation: Some(
                ScalarNode {
                    value: "https://xtensor.readthedocs.io",
                },
            ),
            license: Some(
                ScalarNode {
                    value: "BSD-3-Clause",
                },
            ),
            license_family: Some(
                ScalarNode {
                    value: "BSD",
                },
            ),
            license_file: Some(
                Scalar(
                    ScalarNode {
                        value: "LICENSE",
                    },
                ),
            ),
            license_url: None,
            summary: Some(
                ScalarNode {
                    value: "The C++ tensor algebra library",
                },
            ),
            description: Some(
                ScalarNode {
                    value: "Multi dimensional arrays with broadcasting and lazy computing",
                },
            ),
            prelink_message: None,
        },
    ),
    extra: Extra {
        node: Some(
            Mapping(
                MappingNode {
                    value: {
                        ScalarNode {
                            value: "recipe-maintainers",
                        }: Sequence(
                            SequenceNode {
                                value: [
                                    Simple(
                                        Scalar(
                                            ScalarNode {
                                                value: "some-maintainer",
                                            },
                                        ),
                                    ),
                                ],
                            },
                        ),
                    },
                },
            ),
        ),
    },
}
